### PR TITLE
Move several vector intrinsics out of experimental namespace

### PR DESCRIPTION
**Pull request by mgabka**
This patch is moving out following intrinsics:
* vector.interleave2/deinterleave2
* vector.reverse
* vector.splice

from the experimental namespace.

All these intrinsics exist in LLVM for more than a year now, and are widely used, so should not be considered as experimental.

### CODE REVIEW

**Comment by topperc** [comment]
> vector.revert

I think should be vector.reverse?

**Comment by topperc** [comment]
> All these intrinsics exist in LLVM for more than a year now, and are widly used, so should not be considered as experimental.

widly -> widely

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 1]
```diff
@@ -1099,6 +1099,25 @@ static bool upgradeIntrinsicFunction1(Function *F, Function *&NewFn,
...
+               .StartsWith("interleave2.", Intrinsic::vector_interleave2)
+               .StartsWith("deinterleave2.", Intrinsic::vector_deinterleave2)
+               .Default(Intrinsic::not_intrinsic);
+      if (ID != Intrinsic::not_intrinsic) {
+        const auto *FT = F->getFunctionType();
+        rename(F);
+        if (ID == Intrinsic::vector_interleave2)
+          NewFn = Intrinsic::getDeclaration(F->getParent(), ID,
+                                            FT->getReturnType());
+        else {
```
Comment (topperc) [inline llvm/lib/IR/AutoUpgrade.cpp:1114]: Please be consistent about the use of curly braces between `if` and `else`

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -18779,13 +18779,13 @@ This is an overloaded intrinsic.
...
+      declare <vscale x 4 x i32> @llvm.vector.reverse.nxv4i32(<vscale x 4 x i32> %a)
 
 Overview:
 """""""""
 
-The '``llvm.experimental.vector.reverse.*``' intrinsics reverse a vector.
+The '``llvm.vector.reverse.*``' intrinsics reverse a vector.
 The intrinsic takes a single vector and returns a vector of matching type but
 with the original lane order reversed. These intrinsics work for both fixed
 and scalable vectors. While this intrinsic is marked as experimental the
```
Comment (tschuett) [inline llvm/docs/LangRef.rst:18791]: `this intrinsic is marked as experimental`

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -18914,13 +18914,13 @@ This is an overloaded intrinsic.
...
-      declare <2 x double> @llvm.experimental.vector.splice.v2f64(<2 x double> %vec1, <2 x double> %vec2, i32 %imm)
-      declare <vscale x 4 x i32> @llvm.experimental.vector.splice.nxv4i32(<vscale x 4 x i32> %vec1, <vscale x 4 x i32> %vec2, i32 %imm)
+      declare <2 x double> @llvm.vector.splice.v2f64(<2 x double> %vec1, <2 x double> %vec2, i32 %imm)
+      declare <vscale x 4 x i32> @llvm.vector.splice.nxv4i32(<vscale x 4 x i32> %vec1, <vscale x 4 x i32> %vec2, i32 %imm)
 
 Overview:
 """""""""
 
-The '``llvm.experimental.vector.splice.*``' intrinsics construct a vector by
+The '``llvm.vector.splice.*``' intrinsics construct a vector by
```
Comment (tschuett) [inline llvm/docs/LangRef.rst:18923]: `While this intrinsicis marked as experimental,`

File: llvm/lib/IR/AutoUpgrade.cpp (hunk 1 above)
Comment (mgabka) [reply llvm/lib/IR/AutoUpgrade.cpp]: Thanks for pointing this out, decided to merge this with the similar section for insert/extract.

**Comment by mgabka** [comment]
> > vector.revert
> 
> I think should be vector.reverse?

Will fix the commit message when merging the PR.

**Comment by arsenm** [comment]
> I don't know what the policy is for promoting intrinsics from experimental to first-class or if it's documented anywhere (?), but I would expect this to be accompanied with an RFC / announcement on Discourse.

I don't remember any intrinsic ever making the move out of experimental. It seems to be a permanent prefix

**Comment by nikic** [comment]
> > I don't know what the policy is for promoting intrinsics from experimental to first-class or if it's documented anywhere (?), but I would expect this to be accompanied with an RFC / announcement on Discourse.
> 
> I don't remember any intrinsic ever making the move out of experimental. It seems to be a permanent prefix

At least the vector.reduce intrinsics were moved out of the experimental namespace.

**Comment by mgabka** [comment]
I was actually following this patch https://reviews.llvm.org/D127976 which promoted vector.insert/extract and there was no RFC for it.
My understanding is that these intrinsics are so heavily used in LLVM now so there is no need to call them experimental, moreover the auto upgrade ensures that the "experimental" name is still honoured.

My plan was to promote stepvector as well in separate PR.

I am happy to raise an RFC if you feel that it is worth it.

**Comment by c-rhodes** [comment]
> I was actually following this patch https://reviews.llvm.org/D127976 which promoted vector.insert/extract and there was no RFC for it. My understanding is that these intrinsics are so heavily used in LLVM now so there is no need to call them experimental, moreover the auto upgrade ensures that the "experimental" name is still honoured.
> 
> My plan was to promote stepvector as well in separate PR.

👍

> I am happy to raise an RFC if you feel that it is worth it.

The reduction intrinsics were accompanied with an RFC:
https://lists.llvm.org/pipermail/llvm-dev/2020-April/140729.html

I think at least an announcement with a link to the patch for visibility would be worthwhile.

**Comment by mgabka** [comment]
Rebased on newer LLVM + forcing to re-run the testing which previously failed on some not related issues.

**Comment by c-rhodes** [comment]
> Rebased on newer LLVM + forcing to re-run the testing which previously failed on some not related issues.

Not sure if you missed my previous comment but you need to update the `interleave2` intrinsic in MLIR. MLIR doesn't build with this patch which you can see in pre-commit:
> _bk;t=1713985770943In file included from /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-zw2hg-1/llvm-project/github-pull-requests/mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMIRToLLVMTranslation.cpp:39:
_bk;t=1713985770943[1m/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-zw2hg-1/llvm-project/github-pull-requests/build/tools/mlir/include/mlir/Dialect/LLVMIR/LLVMConvertibleLLVMIRIntrinsics.inc:156:18: [0m[0;1;31merror: [0m[1mno member named 'experimental_vector_interleave2' in namespace 'llvm::Intrinsic'[0m
_bk;t=1713985770943llvm::Intrinsic::experimental_vector_interleave2,

**Comment by mgabka** [comment]
@c-rhodes  the remaining failure is  MLIR :: Conversion/TosaToTensor/tosa-to-tensor.mlir but it also fails without my changes.

File: llvm/test/Bitcode/upgrade-vector-interleave2-deinterleave2-intrinsics.ll [hunk 4]
```diff
@@ -0,0 +1,46 @@
+; RUN: opt -S < %s | FileCheck %s
```
Comment (c-rhodes) [inline llvm/test/Bitcode/upgrade-vector-interleave2-deinterleave2-intrinsics.ll:1]: this `opt` line can be removed? Likewise for other upgrade tests.

**Comment by MacDue** [comment]
> @c-rhodes the remaining failure is MLIR :: Conversion/TosaToTensor/tosa-to-tensor.mlir but it also fails without my changes.

Very likely unrelated we only added `interleave2` for use in ArmSME. It's not used within high-level dialects like tosa & tensor.

**Review by c-rhodes — APPROVED** [review]
one final comment, otherwise LGTM cheers.

File: llvm/test/Bitcode/upgrade-vector-interleave2-deinterleave2-intrinsics.ll (hunk 4 above)
Comment (c-rhodes) [reply llvm/test/Bitcode/upgrade-vector-interleave2-deinterleave2-intrinsics.ll:1]: ah ok, that makes sense 👍

**Review by paulwalker-arm — APPROVED** [review]

**Comment by mgabka** [comment]
I needed to adjust one more test after another rebase:
llvm/test/Transforms/LoopVectorize/AArch64/reduction-recurrence-costs-sve.ll 

I am merging it know without waiting for the pre commit validation, otrherwise will most likely need to update another tests etc.

**Pull request merged**